plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/Inline.hs

```diff
@@ -407,3 +432,124 @@ maybeAddTySubst tn rhs = do
         modify' (extendType tn rhs)
         pure Nothing
     else pure $ Just rhs
+
+applyAndBetaReduce ::
+  forall tyname name uni fun ann.
+  (InliningConstraints tyname name uni fun) =>
+  -- | The rhs of the variable, should have been renamed already
+  Term tyname name uni fun ann ->
+  -- | The arguments, already processed
+  AppContext tyname name uni fun ann ->
+  InlineM tyname name uni fun ann (Term tyname name uni fun ann)
+applyAndBetaReduce rhs args = do
+  info <- ask
+  -- We run the beta and inlining passes for `len args` times, since sometimes the args
+  -- can only be beta-reduced one by one. An example is
+  -- `(\x y -> f x y) arg_x arg_y`, where `arg_x` and `arg_y` are effectful terms.
+  -- It needs two beta+inlining passes to reduce to `f arg_x arg_y`: the first pass
+  -- inlines `y` and the second pass `x`.
+  --
+  -- Note that the inlining passes we run here only run unconditional inlining.
+  -- Technically it can be helpful to also run callsite inlining, but it rarely
+  -- makes a difference and leads to exponential worst-case behavior.
+  foldl'
+    (>=>)
+    pure
+    ( replicate
+        (len args)
+        ( pure . Beta.beta
+            >=> inline'
```

Comment:
`inline'` does still redundantly compute the term usages. 

I'm not sure why we can't just call `processTerm` here? We're *in* `InlineM`, so we can just do that directly and not define `inline'` at all.

Reply:
In fact, because the RHS is renamed, we'll also need to re-calculate the `VarsInfo` (which I'm not doing currently), and merge it with the original `VarsInfo`. I'll update it.

Reply:
Ouch. If we're going to do that, you could compute it locally, merge it with the ones in the context using `local`, and then just call `processTerm`. I think that would be neater IMO.

Reply:
> A further observation

Grr, this is not true because of the inliner state, which is kept in `StateT`, so if we ran `processTerm`, we'd add mappings for stuff we inline to that, and they would _persist_ and potentially be used later where they shouldn't be. So we would need to either rename or only locally modify the state.